source/shared/cpp/ObjectModel/ColumnSet.cpp

```diff
@@ -47,11 +47,84 @@ std::shared_ptr<BaseCardElement> ColumnSetParser::Deserialize(ParseContext& cont
 
 void ColumnSet::DeserializeChildren(ParseContext& context, const Json::Value& value)
-{
-    // Parse Columns
-    auto cardElements =
-        ParseUtil::GetElementCollectionOfSingleType<Column>(context, value, AdaptiveCardSchemaKey::Columns, Column::Deserialize, false);
-    m_columns = std::move(cardElements);
+{ 
+    auto elementArray = ParseUtil::GetArray(value, AdaptiveCardSchemaKey::Columns, false);
+
+    std::vector<std::shared_ptr<Column>> elements;
+    if (elementArray.empty())
+    {
+        m_columns = elements;
+        return;
+    }
+
+    const size_t elemSize = elementArray.size();
+    const size_t endElemIndex = elemSize - 1;
+    elements.reserve(elemSize);
+
+    size_t currentIndex = 0;
+
+    // Child's bleed get more restricted as we walk down the tree
+    // until padding is reset
+    // padding get rested when padding is added to a collection type element
+    // only left child LC or right child RC have access to the padding of their parents
+    // but if parent also has restriction such as BleedToLeft and BleedToRight,
+    // only child that has same restriction get bleed property
+    // for example, if inherited bleed state was BleedToLeft, only LC gets the bleed,
+    // the rest of children cannot bleed.
+    ContainerBleedDirection previousBleedState = context.GetBleedDirection();
+
+    if (elemSize == 1)
+    {
+        // items in signle column has no restriction
+        context.SetBleedDirection(previousBleedState);
+    }
+
+    // Deserialize every element in the array
+    for (const Json::Value& curJsonValue : elementArray)
+    {
+        // processing the cases where parent's bleed state is not restricted
+        if (elemSize != 1)
+        {
+            if (currentIndex == 0)
+            {
+                if (previousBleedState == ContainerBleedDirection::BleedToBothEdges)
+                {
+                    context.SetBleedDirection(ContainerBleedDirection::BleedToLeading);
+                }
+                else if (previousBleedState != ContainerBleedDirection::BleedToLeading)
+                {
+                    context.SetBleedDirection(ContainerBleedDirection::BleedRestricted);
+                }
+            }
+            else if (currentIndex == endElemIndex)
+            {
+                if (previousBleedState == ContainerBleedDirection::BleedToBothEdges)
+                {
+                    context.SetBleedDirection(ContainerBleedDirection::BleedToTrailing);
+                }
+                else if (previousBleedState != ContainerBleedDirection::BleedToTrailing)
+                {
+                    context.SetBleedDirection(ContainerBleedDirection::BleedRestricted);
+                }
+            } else
+            {
+                context.SetBleedDirection(ContainerBleedDirection::BleedRestricted);
+            }
+        }
+
+        InternalId::Next();        
+        // Parse the element
+        auto el = Column::Deserialize(context, curJsonValue);
```

Comment:
I think part of the problem here is that we shouldn't hardcode the assumption that everything under the `columns` key is, in fact, a column. there could be a custom element in the json. we should be going through the element parser registration to parse. if we do that, we shouldn't have to manually move the internal id.

Reply:
that sounds good, but we don't have column parser registered. should we register it?

---
In reply to: [268278405](https://github.com/Microsoft/AdaptiveCards/pull/2551#discussion_r268278405) [](ancestors = 268278405)